src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2166,36 +2222,346 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
 
                     // First, check if we got absolutely nothing.  If so, then 
                     // We need to consume a bad member and try again.
-                    if (explicitInterfaceOpt == null && identifierOrThisOpt == null && typeParameterListOpt == null)
+                    if (IsNoneOrIncompleteMember(parentKind, attributes, modifiers, type, explicitInterfaceOpt, identifierOrThisOpt, typeParameterListOpt, out result))
                     {
-                        if (attributes.Count == 0 && modifiers.Count == 0 && type.IsMissing && !sawRef)
-                        {
-                            // we haven't advanced, the caller needs to consume the tokens ahead
-                            return null;
-                        }
+                        return result;
+                    }
 
-                        var incompleteMember = _syntaxFactory.IncompleteMember(attributes, modifiers.ToList(), type.IsMissing ? null : type);
-                        if (incompleteMember.ContainsDiagnostics)
-                        {
-                            return incompleteMember;
-                        }
-                        else if (parentKind == SyntaxKind.NamespaceDeclaration ||
-                                 parentKind == SyntaxKind.CompilationUnit && !IsScript)
+                    // If the modifiers did not include "async", and the type we got was "async", and there was an
+                    // error in the identifier or its type parameters, then the user is probably in the midst of typing
+                    // an async method.  In that case we reconsider "async" to be a modifier, and treat the identifier
+                    // (with the type parameters) as the type (with type arguments).  Then we go back to looking for
+                    // the member name again.
+                    // For example, if we get
+                    //     async Task<
+                    // then we want async to be a modifier and Task<MISSING> to be a type.
+                    if (ReconsideredTypeAsAsyncModifier(ref modifiers, ref type, ref afterTypeResetPoint, ref explicitInterfaceOpt, ref identifierOrThisOpt, ref typeParameterListOpt))
+                    {
+                        goto parse_member_name;
+                    }
+
+                    Debug.Assert(identifierOrThisOpt != null);
+
+                    // check availability of readonly members feature for indexers, properties and methods
+                    CheckForVersionSpecificModifiers(modifiers, SyntaxKind.ReadOnlyKeyword, MessageID.IDS_FeatureReadOnlyMembers);
```

Comment:
It feels odd to potentially issue LangVersion diagnostics for readonly members here since it doesn't seem like such members are ever allowed at the top level.

Reply:
> It feels odd to potentially issue LangVersion diagnostics for readonly members here since it doesn't seem like such members are ever allowed at the top level.

The more errors we can report for misplaced declaration, the better, IMHO.

---
In reply to: [382868202](https://github.com/dotnet/roslyn/pull/41706#discussion_r382868202) [](ancestors = 382868202)